wordpress-mobile/WordPress-iOS
# Allow for discriminating between beta and stable when building releases on Buildkite

This builds on top of #17891 to address the beta vs stable issue that I didn't notice in previous code reviews. See https://github.com/wordpress-mobile/WordPress-iOS/pull/17891#pullrequestreview-872866603.

The "Rationale" section below is quite verbose. The code should explain itself, but I thought of adding it for extra details. Before diving into it, I'll drop the testing details.

## To Test

The difference between beta and stable is in generating GitHub releases, which can happen only if the ASC submission succeeds. But the ASC submission cannot succeed unless we change the build number, and we don't want to do that.

That's to say, the only way to be 100% sure this work is to run it "for real", after it's been merged.

I attempted to test it as far as it would go by hacking the build in a way to verify my changes behaved as planned. See details in https://github.com/wordpress-mobile/WordPress-iOS/pull/17892.

## Rationale

Before explaining my approach, here's some context on how release builds work on Buildkite. It's actually pretty similar to how they worked on CircleCI, but worth stating here:

- When a release manager runs a lane that should result in a new build, such as `complete_code_freeze` or `finalize_release`...
- The lane internally calls either `trigger_beta_build` or `trigger_release_build`...
- Which in turn call the new `buildkite_trigger_build` to run the pipeline described in `release-builds.yml`

The option I chose to differentiate between a beta release and a stable one was to call Buildkite with environment variable that is then converted to a call parameter for the command script, which in turn converts it into the appropriate `beta_release:true|false` for the build.

The other option I though of was to had two pipeline files for beta and stable, and have `trigger_(beta|release)_build` call the appropriate one. I decided against this option because it would have resulted in too much duplication. Besides it would have either still used the script parameter converted to `beta_release` argument option from the current implementation, or have even more duplication with a dedicated command script for the beta release.

You might also notice that I edited only the WordPress step, not the Jetpack one. That's because the only current effect of differentiating between beta and stable is in the resulting GitHub release being a pre-release or not. Since only the WordPress step generates the GitHub release, there was no need for the Jetpack one to change, too. Besides, the Jetpack build lane doesn't even expect a beta parameter:

https://github.com/wordpress-mobile/WordPress-iOS/blob/9fc374b7ae500881029c07191ca5a0a8eda47bf0/fastlane/Jetpack-Fastfile#L90-L118

## Regression Notes
1. Potential unintended areas of impact
N.A.


2. What I did to test those areas of impact (or what existing automated tests I relied on)
N.A.

3. What automated tests I added (or what prevented me from doing so)
N.A.

---
